dartagnan/src/main/java/com/dat3m/dartagnan/analysis/graphRefinement/graphs/eventGraph/binary/CompositionGraph.java

```diff
@@ -0,0 +1,87 @@
+package com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.binary;
+
+import com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.EventGraph;
+import com.dat3m.dartagnan.analysis.graphRefinement.graphs.eventGraph.utils.MaterializedGraph;
+import com.dat3m.dartagnan.analysis.graphRefinement.util.GraphVisitor;
+import com.dat3m.dartagnan.utils.collections.SetUtil;
+import com.dat3m.dartagnan.utils.timeable.Timestamp;
+import com.dat3m.dartagnan.verification.model.Edge;
+import com.dat3m.dartagnan.verification.model.ExecutionModel;
+
+import java.util.*;
+
+public class CompositionGraph extends MaterializedGraph {
+
+    private final EventGraph first;
+    private final EventGraph second;
+
+    @Override
+    public List<? extends EventGraph> getDependencies() {
+        return Arrays.asList(first, second);
+    }
+
+    public EventGraph getFirst() { return first; }
+    public EventGraph getSecond() { return second; }
+
```

Comment:
All this seems to be common of `CompositionGraph`, `DifferenceGraph`, `MatUnionGraph` and `MatIntersectionGraph`. Wouldn't it be better to have a common abstract class `BinaryGraph` from which they get all this by default?

Reply:
In the beginning there indeed was a ```BinaryGraph``` base class. However, this conflicts with using ```MateralizedGraph``` as a base class since multiple inheritance is not allowed. And ```BinaryGraph``` and ```MateralizedGraph``` have nothing in common. From the perspective of code reusage, there is more value in inheriting from ```MateralizedGraph```. Furthermore, if we eventually go to n-ary Relations /Graphs, having ```BinaryGraph``` is just an obstacle so I removed it altogether.